specs/clipboard-api/clipboard.idl

```diff
@@ -0,0 +1,426 @@
+{
+  [contractversion(1.0)]
+  apicontract ClipboardContract { };
+
+  delegate DataProviderHandler;
+  runtimeclass Clipboard;
+  runtimeclass ClipboardContentOptions;
+  runtimeclass ClipboardHistoryItem;
+  runtimeclass ClipboardHistoryItemsResult;
+  runtimeclass DataPackage;
+  runtimeclass DataPackagePropertySet;
+  runtimeclass DataPackagePropertySetView;
+  runtimeclass DataPackageView;
+  runtimeclass DataProviderDeferral;
+  runtimeclass DataProviderRequest;
+  runtimeclass HtmlFormatHelper;
+  runtimeclass OperationCompletedEventArgs;
+  runtimeclass StandardDataFormats;
+}
+
+// Generic instantiations
+namespace Microsoft.ProjectReunion.ApplicationModel
+{
+  declare
+  {
+    interface Windows.Foundation.Collections.IIterable<ClipboardHistoryItem*>;
+    interface Windows.Foundation.Collections.IIterator<ClipboardHistoryItem*>;
+    interface Windows.Foundation.Collections.IVectorView<ClipboardHistoryItem*>;
+    interface Windows.Foundation.IAsyncOperation<ClipboardHistoryItemsResult*>;
+    interface Windows.Foundation.IAsyncOperation<DataPackage*>;
+    interface Windows.Foundation.IAsyncOperation<DataPackageOperation>;
+    interface Windows.Foundation.TypedEventHandler<DataPackage*, IInspectable*>;
+    interface Windows.Foundation.TypedEventHandler<DataPackage*, OperationCompletedEventArgs*>;
+  }
+}
+
+// Type definitions
+namespace Microsoft.ProjectReunion.ApplicationModel
+{
+
+  // FIXME: Figure out how to declare these enums without causing definition conflicts
+  // between the enum constants and the same-named ones in
+  // Windows.ApplicationModel.DataTransfer
+
+  [contract(ClipboardContract, 1.0)]
+  enum ClipboardHistoryItemsResultStatus
+  {
+    aaaSuccess                  = 0,
+    aaaAccessDenied             = 1,
+    aaaClipboardHistoryDisabled = 2,
+  };
+
+  [contract(ClipboardContract, 1.0)]
+  [flags]
+  enum DataPackageOperation
+  {
+    // TODO: combine Copy and Move?
+    aaaNone = 0x0,
+    aaaCopy = 0x1,
+    aaaMove = 0x2,
+    aaaLink = 0x4,
+  };
+
+  [contract(ClipboardContract, 1.0)]
+  enum SetHistoryItemAsContentStatus
+  {
+    aaaSuccess      = 0,
+    aaaAccessDenied = 1,
+    aaaItemDeleted  = 2,
+  };
+
+  /// <summary>
+  /// Prototype for functions that will be called to render a delay-rendered
+  /// data format in a <see cref="DataPackage"/>.
+  /// </summary>
+  [contract(ClipboardContract, 1.0)]
+  delegate void DataProviderHandler(DataProviderRequest request);
+
+  /// <summary>
+  /// Top-level API for reading, writing, and monitoring the system clipboard,
+  /// as well as clipboard history, if available on the current Windows version
+  /// and enabled by the user. Clipboard history also includes clipboard items
+  /// roamed from your other devices through the cloud.
+  /// </summary>
+  [contract(ClipboardContract, 1.0)]
+  [marshaling_behavior(standard)]
+  static runtimeclass Clipboard
+  {
+    static DataPackageView GetContent();

```

Comment:
I hope we can make this call more robust. It's not rare for this to fail with a set of "known" exceptions and then succeed on a retry. It would be a better developer experience if these retries are at least encapsulated and the call either fails or succeeds. Ideally, of course, if must never fail, as there is no _obvious_ reason for not being able to read out from clipboard (if we forget old Win32 API design and such), even if it means that clipboard lock is forcefully re-taken by the caller.